### PR TITLE
[Example][Refactor] Refactor graphsage multigpu and full-graph example

### PR DESCRIPTION
**Pull request by chang-l**
## Description
This is to address https://github.com/dmlc/dgl/issues/4409, referring to https://github.com/dmlc/dgl/pull/4186 and https://github.com/dmlc/dgl/pull/4160.

For `train_full.py`,  adopted all default settings/options from the old implementation, and made it very similar to `gcn/train.py`.
For `multi_gpu_node_classification.py`, compared to the old one, I added two cmd line options: (1) `--mode` in consistent with its single-gpu implementation `node_classification.py` (tested both modes), (2) `--gpu` to control the # of gpus to run.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Results 

- `train_full.py` (5 runs)

| Dataset       | Test acc. (before) | Mean | Test acc. (after) | Mean
| ------------- | -------| -------| -------| -------
| cora  | [0.812 0.82 0.818 0.811 0.822] | 0.8166 | [0.815 0.806 0.811 0.809 0.82] | 0.8122
| citeseer | [0.710 0.710 0.698 0.704 0.704] | 0.7052 | [0.715 0.702 0.696 0.712 0.705] | 0.7060
| pubmed       | [0.788 0.780 0.782 0.780 0.781] | 0.7822 | [0.782 0.786 0.782 0.782 0.784] | 0.7832

- `multi_gpu_node_classificaiton.py` (3 runs)

| Dataset       | Test acc. (before) | Mean |Test acc. (after) | Mean
| ------------- | -------| -------| -------| -------
| ogbn-product| [0.726 0.720 0.727] | 0.7243 | [0.731 0.727 0.727] | 0.7283

## Concerns

- Accuracy drop in multi-gpu runs
   For `multi_gpu_node_classification.py`, when using 1 GPU training, the result is consistent with the one from `node_classification.py`, which is around 0.770. However, if we use 2GPU or more (e.g., 4) GPUs, the results are dropped to ~0.720 (consistently among >10 runs, or check the above table).  I suspect there is a bug and we need to fix it.

- Poor scaling performance on multi-gpu inference 
  For multi-gpu inference in `multi_gpu_node_classification.py`, after a detailed profiling, it seems the cost of pin/unpin memory scales up and becomes significant when increasing GPUs.

### CODE REVIEW

**Comment by mufeili** [comment]
There is a conflict that needs to be addressed.

**Comment by mufeili** [comment]
While these numbers indeed appear close, it will be better if you can add the mean values.

**Comment by mufeili** [comment]
> Accuracy drop in multi-gpu runs
For multi_gpu_node_classification.py, when using 1 GPU training, the result is consistent with the one from node_classification.py, which is around 0.770. However, if we use 2GPU or more (e.g., 4) GPUs, the results are dropped to ~0.720 (consistently among >10 runs, or check the above table). I suspect there is a bug and we need to fix it.

I think this is quite common. With the use of more GPUs, it's possible that you need to adjust the hyperparameters.

File: examples/pytorch/graphsage/train_full.py [hunk 1]
```diff
@@ -1,171 +1,93 @@
...
-    parser.add_argument("--n-hidden", type=int, default=16,
-                        help="number of hidden gcn units")
-    parser.add_argument("--n-layers", type=int, default=1,
-                        help="number of hidden gcn layers")
-    parser.add_argument("--weight-decay", type=float, default=5e-4,
-                        help="Weight for L2 loss")
-    parser.add_argument("--aggregator-type", type=str, default="gcn",
-                        help="Aggregator type: mean/gcn/pool/lstm")
+    parser.add_argument("--dataset", type=str, default="cora",
+                        help="Dataset name ('cora', 'citeseer', 'pubmed').")
```
Comment (mufeili) [inline examples/pytorch/graphsage/train_full.py:60]: It's strange to have "." at the end.

File: examples/pytorch/graphsage/train_full.py [hunk 2]
```diff
@@ -1,171 +1,93 @@
...
     args = parser.parse_args()
-    print(args)
-
-    main(args)
+    print(f'Training with DGL built-in GraphSage module.')
```
Comment (mufeili) [inline examples/pytorch/graphsage/train_full.py:62]: Ditto

File: examples/pytorch/graphsage/train_full.py [hunk 3]
```diff
@@ -1,171 +1,93 @@
...
-Simple reference implementation of GraphSAGE.
-"""
-import argparse
-import time
-import numpy as np
-import networkx as nx
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
 import dgl
```
Comment (mufeili) [inline examples/pytorch/graphsage/train_full.py:4]: No longer used

File: examples/pytorch/graphsage/train_full.py [hunk 4]
```diff
@@ -1,171 +1,93 @@
...
-
-    # graph preprocess and calculate normalization factor
-    g = dgl.remove_self_loop(g)
-    n_edges = g.number_of_edges()
-    if cuda:
-        g = g.int().to(args.gpu)
+def train(g, features, labels, masks, model):
+    # define train/val samples, loss function and optimizer
+    train_mask = masks[0]
+    val_mask = masks[1]
```
Comment (mufeili) [inline examples/pytorch/graphsage/train_full.py:41]:
```suggestion
    train_mask, val_mask, _ = masks
```

File: examples/pytorch/graphsage/train_full.py [hunk 5]
```diff
@@ -1,171 +1,93 @@
...
+    elif args.dataset == 'pubmed':
+        data = PubmedGraphDataset(transform=transform)
+    else:
+        raise ValueError('Unknown dataset: {}'.format(args.dataset))
+    g = data[0]
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    g = g.int().to(device)
+    features = g.ndata['feat']
+    labels = g.ndata['label']
+    masks = g.ndata['train_mask'], g.ndata['val_mask'], g.ndata['test_mask']
```
Comment (mufeili) [inline examples/pytorch/graphsage/train_full.py:79]: Maybe just do `masks = g.ndata['train_mask'], g.ndata['val_mask']` and pass `g.ndata['test_mask']` directly at L92.

File: examples/pytorch/graphsage/multi_gpu_node_classification.py [hunk 6]
```diff
@@ -1,182 +1,164 @@
...
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
-import torch.distributed as dist
-import torch.distributed.optim
 import torchmetrics.functional as MF
+import torch.distributed as dist
+from torch.nn.parallel import DistributedDataParallel
+import torch.multiprocessing as mp
 import dgl
```
Comment (mufeili) [inline examples/pytorch/graphsage/multi_gpu_node_classification.py:9]: `dgl` no longer used

File: examples/pytorch/graphsage/multi_gpu_node_classification.py [hunk 7]
```diff
@@ -1,182 +1,164 @@
...
-    # Tested with mp.spawn and fork.  Both worked and got 4s per epoch with 4 GPUs
-    # and 3.86s per epoch with 8 GPUs on p2.8x, compared to 5.2s from official examples.
-    import torch.multiprocessing as mp
-    mp.spawn(train, args=(n_procs, graph, num_classes, split_idx), nprocs=n_procs)
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--mode", default='mixed', choices=['mixed', 'puregpu'],
+                        help="Training mode. 'mixed' for CPU-GPU mixed training, "
+                             "'puregpu' for pure-GPU training.")
+    parser.add_argument("--gpu", type=str, default='0',
+                           help="GPU(s) in use. Can be a list of gpu ids for multi-gpu training,"
```
Comment (mufeili) [inline examples/pytorch/graphsage/multi_gpu_node_classification.py:146]: Fix the indent

File: examples/pytorch/graphsage/multi_gpu_node_classification.py [hunk 8]
```diff
@@ -1,182 +1,164 @@
...
+    
+    # load and preprocess dataset
+    print('Loading data')
+    dataset = AsNodePredDataset(DglNodePropPredDataset('ogbn-products'))
+    g = dataset[0]
+    # avoid creating certain graph formats in each sub-process to save momory
+    g.create_formats_()
+    # thread limiting to avoid resource competition
+    os.environ['OMP_NUM_THREADS'] = str(mp.cpu_count() // 2 // nprocs)    
+    dist_data = dataset.num_classes, dataset.train_idx, dataset.val_idx, dataset.test_idx
```
Comment (mufeili) [inline examples/pytorch/graphsage/multi_gpu_node_classification.py:162]: Not sure if `dist_data` is better than `data`

File: examples/pytorch/graphsage/multi_gpu_node_classification.py [hunk 9]
```diff
@@ -1,182 +1,164 @@
...
+    # find corresponding device for my rank
+    device = devices[proc_id]
+    torch.cuda.set_device(device)
+    # initialize process group and unpack data for sub-processes
+    dist.init_process_group(backend="nccl", init_method='tcp://127.0.0.1:12345', world_size=nprocs, rank=proc_id)
+    out_size, train_idx, val_idx, test_idx = dist_data
+    train_idx = train_idx.to(device)
+    val_idx = val_idx.to(device)
+    g = g.to(device if mode == 'puregpu' else 'cpu')
+    # create RGCN model (distributed)
```
Comment (mufeili) [inline examples/pytorch/graphsage/multi_gpu_node_classification.py:129]: GraphSAGE?

File: examples/pytorch/graphsage/multi_gpu_node_classification.py [hunk 10]
```diff
@@ -1,182 +1,164 @@
...
+    in_size = g.ndata['feat'].shape[1]
+    model = SAGE(in_size, 256, out_size).to(device)
+    model = DistributedDataParallel(model, device_ids=[device], output_device=device)
+    # training + testing
+    use_uva = (mode == 'mixed')
+    train(proc_id, nprocs, device, g, train_idx, val_idx, model, use_uva)
+    layerwise_infer(proc_id, device, g, test_idx, model, use_uva, batch_size=2**16)
```
Comment (mufeili) [inline examples/pytorch/graphsage/multi_gpu_node_classification.py:136]: I'll prefer directly setting the default value of `batch_size` in defining `layerwise_infer`.

**Review by mufeili — APPROVED** [review]
I'm good other than some minor comments.

File: examples/pytorch/graphsage/train_full.py (hunk 1 above)
Comment (chang-l) [reply examples/pytorch/graphsage/train_full.py]: Removed.

File: examples/pytorch/graphsage/train_full.py (hunk 2 above)
Comment (chang-l) [reply examples/pytorch/graphsage/train_full.py]: Removed.

File: examples/pytorch/graphsage/train_full.py (hunk 3 above)
Comment (chang-l) [reply examples/pytorch/graphsage/train_full.py]: Thanks. I fixed.

File: examples/pytorch/graphsage/train_full.py (hunk 4 above)
Comment (chang-l) [reply examples/pytorch/graphsage/train_full.py]: Sure. I updated accordingly.

File: examples/pytorch/graphsage/multi_gpu_node_classification.py (hunk 7 above)
Comment (chang-l) [reply examples/pytorch/graphsage/multi_gpu_node_classification.py]: Thanks. Fixed.

File: examples/pytorch/graphsage/multi_gpu_node_classification.py (hunk 8 above)
Comment (chang-l) [reply examples/pytorch/graphsage/multi_gpu_node_classification.py]: Right. changed `dist_data` --> `data`

File: examples/pytorch/graphsage/multi_gpu_node_classification.py (hunk 10 above)
Comment (chang-l) [reply examples/pytorch/graphsage/multi_gpu_node_classification.py]: Updated accordingly

**Comment by chang-l** [comment]
> While these numbers indeed appear close, it will be better if you can add the mean values.

Sure. I will add mean values in future PR description.

**Comment by mufeili** [comment]
@chang-l You can merge the PR once ready.

**Comment by chang-l** [comment]
Sync with current master, but I don't think I have write access... Can you please help merge it? @mufeili  Thanks.

**Comment by yaox12** [comment]
Seems the CI is down.

**Comment by mufeili** [comment]
> Sync with current master, but I don't think I have write access... Can you please help merge it? @mufeili Thanks.

Got it. I'll do that when CI is back.

**Comment by Rhett-Ying** [comment]
Commit ID: 47d1d96f842bd7e3cbe7a64eee18c87dc2095d2c

Build ID: 10

Status: ✅ CI test succeeded 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4430/10/10/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4430/10/10/logs/cireport.log)
